Add SessionPools and Sessions

yitaopan opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/20296335/35140fb9-639e-4e44-ac4b-38b1a20dddf9)

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- For help with Step 2 / "ARM review", see https://aka.ms/azsdk/pr-arm-review.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [ ] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [x] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

yitaopan committed:
Add SessionPools.json and Sessions.json

yitaopan committed:
Prettier

TimLovellSmith commented:
Please complete PR description, Purpose of PR, due diligence etc, fix Prettier failures, and review LintDiff for any actionable errors

Commit:
Update Sessions Api, move count from query parameter to body

yitaopan committed:
Update Sessions Api, create a specific type for updateEnvVars

yitaopan committed:
Merge, resolve conflicts

yitaopan committed:
Prettier

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Test

yitaopan committed:
Test

yitaopan committed:
Test

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Add minWarmPoolSize and update coolDownPeriod

yitaopan committed:
Fix - add minwarmpoolsize for patch

yitaopan committed:
Remove MinWarmPoolSize and add ScaleConfiguration

bowen5 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,683 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.App/sessionpools": {
```
Should this be `seesionPools`? Then we don't need to change cSpell.json.

yitaopan replied:
Yes, previously in the CP code I saw it was Microsoft.Apps/sessionpools so I used this, but sessionPools makes more sense, I have fix this.

yitaopan committed:
sessionpool to sessionPool

yitaopan committed:
Move maxConcurrentSessions into scaleConfiguration

yitaopan committed:
Update coolDownPeriod to cooldownPeriodInSeconds

yitaopan committed:
Merge branch 'release-app-Microsoft.App-2024-02-02-preview' into 20240202p-session-sessionpool

bowen5 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/examples/SessionPools_CreateOrUpdate.json:
```diff
@@ -0,0 +1,140 @@
+  "parameters": {
+    "subscriptionId": "34adfa4f-cedf-4dc0-ba29-b6d1a69ab345",
+    "resourceGroupName": "rg",
+    "sessionPoolName": "testsessionpool0",
```
>testsessionpool0 [](http://example.com/codeflow?start=24&length=16)

should this be `testsessionpool` considering L50, L89, etc.

yitaopan replied:
Thanks, fixed

yitaopan committed:
Update example sessionPoolName

bowen5 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/examples/SessionPools_Patch.json:
```diff
@@ -0,0 +1,95 @@
+        }
+      }
+    },
+    "202": {
```
>202 [](http://example.com/codeflow?start=5&length=3)

Do we need to return the payload body, according to [https://github.com/cloud-and-ai-microsoft/resource-provider-contract/blob/master/v1.0/resource-api-reference.md#patch-resource](https://github.com/cloud-and-ai-microsoft/resource-provider-contract/blob/master/v1.0/resource-api-reference.md#patch-resource)?

yitaopan replied:
I think yes, but I referenced other resources like managedEnvironment, container app, container app job, and they don't have body for patch 202 response, so I followed this pattern for now.

bowen5 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/examples/Sessions_UpdateEnvironmentVariables.json:
```diff
@@ -0,0 +1,48 @@
+            {
+              "name": "name-1-1",
+              "value": "value-1-1",
+              "secretRef": "secretRef-1-1"
```
`value` and `secretRef` should be mutual exclusive I suppose?

yitaopan replied:
Thanks, fixed

yitaopan committed:
Fix Sessions UpdateEnvVar example

yitaopan committed:
Fix SessionPool example resources

ms-henglu commented:
Please check and fix the lint diff errors: https://github.com/Azure/azure-rest-api-specs/pull/28047/checks?check_run_id=23096245701

yitaopan committed:
Merge branch 'release-app-Microsoft.App-2024-02-02-preview' into 20240202p-session-sessionpool

yitaopan committed:
Remove update session env var api

yitaopan committed:
Refactor Session apis

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Update to customContainerTemplate

xwang971 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,678 @@
+          }
+        }
+      },
+      "patch": {
```
We don't have this method from CP right? If we don't have plan to support patch at this version from CP, we should remove this, otherwise the SDK would return 404 when call this method.

yitaopan replied:
Thanks, I have added the Patch method in that Update SessionPool Apis align with Swagger CP PR (still under review). Since Patch is required for a tracked arm resource https://github.com/Azure/azure-openapi-validator/blob/main/docs/tracked-resource-patch-operation.md

BlackRider97 replied:
+1 Not sure about PATCH operation. Please double confirm this one.

TimLovellSmith replied:
You support patch for at least tags right? Since its required.

yitaopan committed:
Add RunningState and RunningStateDetails for SessionPool

yitaopan committed:
Make RunningState string

yitaopan committed:
Add enableEgress to sessionPool

yitaopan committed:
Remove SessionAuthConfiguration from sessions to sessionPools

yitaopan committed:
Fix and add node count for session pool

yitaopan committed:
Fix patch

yitaopan committed:
Fix typo

yitaopan committed:
Fix typo

yitaopan committed:
Merge branch 'release-app-Microsoft.App-2024-02-02-preview' into 20240202p-session-sessionpool

yitaopan commented:
Would like to apply for suppression for the linter errors about error schema thanks

raosuhas commented:
> Would like to apply for suppression for the linter errors about error schema thanks

What is the reason for the suppression ?  The response must follow the guidelines as explained here : https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content

raosuhas commented:
You still have a large number of lint warnings here : https://github.com/Azure/azure-rest-api-specs/pull/28047/checks?check_run_id=23988239462 could you also please fix those ?

yitaopan committed:
Update to v5 commontypes

yitaopan committed:
Add array identifiers

yitaopan committed:
Fix more descriptions

yitaopan commented:
> You still have a large number of lint warnings here : https://github.com/Azure/azure-rest-api-specs/pull/28047/checks?check_run_id=23988239462 could you also please fix those ?

Thanks, I have fixed most of the lint warnings.
Could we apply for suppression for the rest of them for I have checked to be necessary, thanks.
For the error response, this is using the same error response as other apis

raosuhas commented:
> > You still have a large number of lint warnings here : https://github.com/Azure/azure-rest-api-specs/pull/28047/checks?check_run_id=23988239462 could you also please fix those ?
> 
> Thanks, I have fixed most of the lint warnings. Could we apply for suppression for the rest of them for I have checked to be necessary, thanks. For the error response, this is using the same error response as other apis

Thanks. I think you still to fix this one though : 

<html>
<body>
<!--StartFragment-->
ListInOperationName | Since operation response has model definition in array type, it should be of the form '_list'.Location: Microsoft.App/preview/2024-02-02-preview/Sessions.json#L37
-- | --


<!--EndFragment-->
</body>
</html>[ListInOperationName](https://github.com/Azure/azure-openapi-validator/blob/main/docs/list-in-operation-name.md)	Since operation response has model definition in array type, it should be of the form '_list'.
Location: [Microsoft.App/preview/2024-02-02-preview/Sessions.json#L37](https://github.com/Azure/azure-rest-api-specs/blob/7c19ebfcdde0b46b333e83510b132412a9764296/specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json#L37)

yitaopan committed:
Modify generate sessions operationId

yitaopan committed:
Modify generate sessions operationId

yitaopan commented:
> > > You still have a large number of lint warnings here : https://github.com/Azure/azure-rest-api-specs/pull/28047/checks?check_run_id=23988239462 could you also please fix those ?
> > 
> > 
> > Thanks, I have fixed most of the lint warnings. Could we apply for suppression for the rest of them for I have checked to be necessary, thanks. For the error response, this is using the same error response as other apis
> 
> THanks can you fix this one as well . The other warnings are ok to ignore if you have already reviewed them ![image](https://private-user-images.githubusercontent.com/11067581/324031576-689c1f7c-acc7-4e93-8e73-385e43429336.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1NDQ4MjMsIm5iZiI6MTcxMzU0NDUyMywicGF0aCI6Ii8xMTA2NzU4MS8zMjQwMzE1NzYtNjg5YzFmN2MtYWNjNy00ZTkzLThlNzMtMzg1ZTQzNDI5MzM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDE2MzUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NjQwMzAzNGE0YTIwYWQwNDY1ZjE5MDEzMjkzNWU1ZTQyY2E0MmU2MjQ2ZWI2ZmU3MTAyYjNkMTdiMmIyZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FLFyX3db_qj3RP3lksLa9c7bjSw_rty0QQ97vOhcgsU)
> 
> For the error schema , we expect all ARM apis to confirm to the same error schema that I shared in the guidelines above. How is your response different from that schema and why cant you use the definition from common types ?

Hi, I have fixed the Lint warning you mentioned. Could you please take another look, thanks!

For the error schema, the APIs in this PR (SessionPools and Sessions), are using the `DefaultErrorResponse` defined in the `CommonDefinitions` file in the folder of this service, which is the same as other APIs, like ContainerApps, ManagedEnvironments, JavaComponents, etc. 
If changing this is necessary, I think we might need to change all these APIs as well, which could be outside the scope of this PR, maybe we can do this in a future PR.

raosuhas commented:
> > > > You still have a large number of lint warnings here : https://github.com/Azure/azure-rest-api-specs/pull/28047/checks?check_run_id=23988239462 could you also please fix those ?
> > > 
> > > 
> > > Thanks, I have fixed most of the lint warnings. Could we apply for suppression for the rest of them for I have checked to be necessary, thanks. For the error response, this is using the same error response as other apis
> > 
> > 
> > THanks can you fix this one as well . The other warnings are ok to ignore if you have already reviewed them ![image](https://private-user-images.githubusercontent.com/11067581/324031576-689c1f7c-acc7-4e93-8e73-385e43429336.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1NDQ4MjMsIm5iZiI6MTcxMzU0NDUyMywicGF0aCI6Ii8xMTA2NzU4MS8zMjQwMzE1NzYtNjg5YzFmN2MtYWNjNy00ZTkzLThlNzMtMzg1ZTQzNDI5MzM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDE2MzUyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NjQwMzAzNGE0YTIwYWQwNDY1ZjE5MDEzMjkzNWU1ZTQyY2E0MmU2MjQ2ZWI2ZmU3MTAyYjNkMTdiMmIyZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FLFyX3db_qj3RP3lksLa9c7bjSw_rty0QQ97vOhcgsU)
> > For the error schema , we expect all ARM apis to confirm to the same error schema that I shared in the guidelines above. How is your response different from that schema and why cant you use the definition from common types ?
> 
> Hi, I have fixed the Lint warning you mentioned. Could you please take another look, thanks!
> 
> For the error schema, the APIs in this PR (SessionPools and Sessions), are using the `DefaultErrorResponse` defined in the `CommonDefinitions` file in the folder of this service, which is the same as other APIs in this service, like ContainerApps, ManagedEnvironments, JavaComponents, etc. If changing this is necessary, I think we might need to change all these APIs as well, which could be outside the scope of this PR, maybe we can do this in a future PR.

Yes can you please take an item. I took a look at your definition and it seems like you are using extra field called innerError : 
https://github.com/Azure/azure-rest-api-specs/blob/eb131efda388677382f60fa7ee731a9e87232f0e/specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/CommonDefinitions.json#L384

This is not really recommended tools built on top of azure will not know to read this field and is not consistent with the rest of azure apis. Is this field being used extensively already ?

For now please follow the suppression process to add suppression for the LRO error

raosuhas commented:
suppression process is described in the PR description here : 
![image](https://github.com/Azure/azure-rest-api-specs/assets/11067581/895987cc-08a3-49a5-9f7c-60e6f7d5cc79)

yitaopan committed:
Modify generate sessions operationId

yitaopan committed:
Add suppression

yitaopan commented:
> suppression process is described in the PR description here : ![image](https://private-user-images.githubusercontent.com/11067581/324092467-895987cc-08a3-49a5-9f7c-60e6f7d5cc79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1ODIwODgsIm5iZiI6MTcxMzU4MTc4OCwicGF0aCI6Ii8xMTA2NzU4MS8zMjQwOTI0NjctODk1OTg3Y2MtMDhhMy00OWE1LTlmN2MtNjBlNmY3ZDVjYzc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIwVDAyNTYyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjNGUzYThjNmJmMjk4Y2I4NTFmM2Y2YTdlNzJkMzBmMjBmMWQ1NTkxYzQ1Mzk3MTdjM2RiMTA5ZTE1ZTYyOGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jRGTG7i_s8rOQgfEO5a37pkk4G59NBckpup2NKQ3Z2E)

Thanks, I have updated the suppressions. And I will discuss the error response with the team.

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+          "format": "date-time"
+        },
+        "connectionString": {
+          "description": "The connection string of the Session.",
```
Since it is sensitive information then we should not return back in client response.  More information: https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#rest-secrets-allowed-in-post-response

Not sure if this is applicable for ARM APIs too

yitaopan replied:
Removed this now

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
if this is async operation then status should be 202  https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md

And if it is not an async operation, then it needs to be under <1 sec completion SLA I guess.

BlackRider97 replied:
@yitaopan I am assuming server implementation is also as per API spec as previously it was mentioned as sync operation in API spec.

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/sessionPools/{sessionPoolName}/sessions": {
```
delete a session API is not needed?

yitaopan replied:
Yes, sessions would be auto deleted.

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,747 @@
+      "patch": {
+        "tags": [
+          "ContainerAppsSessionPools"
+        ],
```
If there is equivalent create API, then have you considered async-operation 202 status code for that?

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+          "type": "string",
+          "format": "date-time"
+        },
+        "generatedTime": {
```
instead of `generatedTime` would you consider `generateAt` as mentioned here 
<img width="374" alt="image" src="https://github.com/Azure/azure-rest-api-specs/assets/4725916/fb192cc2-d300-41f0-8bb6-2ee7e2b0a058">

https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md

yitaopan replied:
name fixed

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,747 @@
+                "modelAsString": true
+              }
+            },
+            "poolManagementType": {
```
Sticking to Kind or Type suffix would be better for consistency? I have Kind suffix in the session.

yitaopan replied:
Updated name

BlackRider97 replied:
So, you are proceeding with `Type` everywhere instead of `Kind` ?

TimLovellSmith replied:
Is this being used as a discriminator? Considered using swagger discriminator notations? x-ms-discriminator-value etc

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,747 @@
+      "description": "Session Pool Private Registry Credentials.",
+      "type": "object",
+      "properties": {
+        "registryServer": {
```
defining as required property would be helpful?

BlackRider97 commented on specification/app/resource-manager/readme.md:
```diff
@@ -65,11 +65,33 @@ input-file:
+    from: Sessions.json
+    reason: |
+      The operationId already using SessionsGeneration.
+  - suppress: MissingSegmentsInNestedResourceListOperation
```
Not sure if these suppression are actually or not. Please re-confirm if it is not actual issue

ramoka178 commented on specification/app/resource-manager/readme.md:
```diff
@@ -65,11 +65,33 @@ input-file:
+  - suppress: PostOperationIdContainsUrlVerb
+    from: SessionPools.json
+    reason: |
+      The operationId already using SessionsGeneration.
```
Dont see any post operation in SessionPools.json. Please remove if not needed.

ramoka178 commented on specification/app/resource-manager/readme.md:
```diff
@@ -65,11 +65,33 @@ input-file:
+  - suppress: PostOperationIdContainsUrlVerb
+    from: Sessions.json
+    reason: |
+      The operationId already using SessionsGeneration.
```
Possible to change operationId ContainerAppsSessionPoolSessionsGeneration_ListGenerated to have the action ?

ramoka178 commented on specification/app/resource-manager/readme.md:
```diff
@@ -65,11 +65,33 @@ input-file:
+  - suppress: EnumInsteadOfBoolean
+    from: SessionPools.json
+    reason: |
+      Boolean is needed for authEnabled and enableEgress
```
>Boolean is needed for authEnabled and enableEgress [](http://example.com/codeflow?start=6&length=50)

Standard guidance is: replace boolean/switch properties with a more meaningful enum whenever possible. 

A boolean will forever have two valid values (true or false). A string enum type is always preferred. Also, properties should always provide better values just than True and False. For example two switches "isTypeA" and "isTypeB" should be replaced with one enum "type": [A, B, DefaultType]. Enums are always a more flexible and future proof option because they allow additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated]. 

Note: do NOT define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its cofusing, and no better than a boolean.

yitaopan committed:
Remove Sessions connectionString

yitaopan committed:
Update operation id for generate sessions

yitaopan committed:
Add SessionPoolNameParameter in commondefinitions

yitaopan committed:
generatedat and expiredat

yitaopan committed:
sessionKind to sessionType

yitaopan committed:
Remove runningState and runningStateDetails for SessionPool

yitaopan committed:
Fix readonly

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Test

yitaopan committed:
Remove two supressions

yitaopan committed:
Change boolean to enum

yitaopan committed:
Fix examples

jijohn14 commented on specification/app/resource-manager/readme.md:
```diff
@@ -65,11 +65,17 @@ input-file:
     from: JavaComponents.json
     reason: |
       Java Component is using componentType as the discriminator. While the discriminator is a required property, this rule prevent it being present in the patch request body.
+  - suppress: LroErrorContent
```
what is the error related to this?

yitaopan replied:
Hi, I removed the suppression temporary, and there would be lint error now https://github.com/Azure/azure-rest-api-specs/pull/28047/checks?check_run_id=24182704181
![image](https://github.com/Azure/azure-rest-api-specs/assets/101094705/dd2517bc-eeb6-46a5-b368-f37b03c94cb4)

yitaopan committed:
Fix examples

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/examples/SessionPools_Delete.json:
```diff
@@ -0,0 +1,16 @@
+  "responses": {
+    "202": {
+      "headers": {
+        "location": "https://management.azure.com/subscriptions/34adfa4f-cedf-4dc0-ba29-b6d1a69ab345/providers/Microsoft.App/locations/eastus/operationResults/amF3YUNvbXBvbmVudHM6OWFiZTM5OGUtY2ZjNi00NGZmLThmODQtNjRiOWJhMTUzZWYy?api-version=2023-08-01-preview"
```
Thanks! fixed

yitaopan committed:
Temp remove suppression

yalixiang commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,737 @@
+              "enum": [
+                "JupyterPython",
+                "FunctionsPython",
+                "CustomContainer"
```
Are we using these, I think last time we discussed that we will use `python-lts` and `node-lts` ? We may need double confirm

yitaopan replied:
Thanks, missed this part, fixed now

yitaopan committed:
Fix SessionPools ContainerType

yalixiang commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,736 @@
+        "sessionAuthConfiguration": {
+          "description": "Auth configuration of the Sessions in this Session Pool.",
+          "$ref": "#/definitions/SessionAuthConfiguration"
+        }
```
Do we need to have an auth configuration here?

yitaopan replied:
Thanks, confirmed that this is removed, removing this now

yitaopan committed:
Remove SessionAuthConfiguration

yitaopan committed:
Add Suppression back

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+          "ContainerAppsSessionPoolSessions"
+        ],
+        "summary": "Generate Sessions from a Session Pool.",
+        "operationId": "ContainerAppsSessionPoolSessionsGeneration_ListGenerated",
```
Thanks, changed to ContainerAppsSessionPoolSessionsGeneration_GenerateSessions

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+            "maxLength": 63
+          },
+          {
+            "name": "name",
```
Added a parameter section

yitaopan committed:
Add constraint to count of generated sessions

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+          "type": "object",
+          "properties": {
+            "count": {
+              "description": "The number of Sessions to generate.",
```
Thanks, constraint is added.

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+          "type": "object",
+          "properties": {
+            "name": {
+              "description": "The name of the Session.",
```
No, this should be returned by the service rather than decided by the users.

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+          "type": "integer",
+          "format": "int32"
+        },
+        "expiryTime": {
```
Readonly properties marked

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,374 @@
+            "modelAsString": true
+          }
+        },
+        "status": {
```
Turned into enums from booleans

yitaopan committed:
Make Session generation async

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Remove pagable

yitaopan committed:
Post to Put

yitaopan committed:
post

yitaopan committed:
post

yitaopan committed:
long running post

yitaopan committed:
Fix

yitaopan committed:
Suppression:

yitaopan commented:
Hi @BlackRider97 @ramoka178, thanks! comments are addressed, could you please take a look again?

ramoka178 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+              "description": "The container type of the Sessions.",
+              "enum": [
+                "python-lts",
+                "node-lts"
```
existing enums 

"JupyterPython",
 "FunctionsPython",
"CustomContainer"

are not required anymore?

yitaopan replied:
Yes, we are changing these to the current one

BlackRider97 replied:
why enum values are not following `Pascal Casing` pattern ? If there are well known values then it should be okay.

ramoka178 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+              "$ref": "#/definitions/DynamicPoolConfiguration"
+            },
+            "customContainerTemplate": {
+              "description": "The custom container configuration if the containerType is node-lts.",
```
>is node-lts. [](http://example.com/codeflow?start=86&length=12)

there are other enums too as in swagger. Can this be generic description ?

yitaopan replied:
Only when the containerType is node-lts, users need to specify customContainerTemplate, and only when the poolType is dynamic, users needs to specify dynamicPoolConfiguration.
May I ask what do you mean by generic description? Thanks

BlackRider97 replied:
In future if we need support for another container type under custom container then it will be confusing to users as description gives the impression that it's only applicable for `node-lts`

ramoka178 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,387 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/sessionPools/{sessionPoolName}/sessions/{name}": {
```
>name [](http://example.com/codeflow?start=137&length=4)

sessionName?

yitaopan replied:
Thanks, updated

ramoka178 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,387 @@
+            "type": "string",
+            "minLength": 3,
+            "maxLength": 63
+          },
```
parameter like SessionPoolNameParameter ?

yitaopan replied:
Sure, previously because this is only used here so I didn't parameterize this, but I agree with you that it would be better to do this, thanks.

ramoka178 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,387 @@
+          "type": "string",
+          "format": "date-time",
+          "readOnly": true
+        },
```
mention the format/timezone of time in description

yitaopan replied:
Fixed

BlackRider97 replied:
Description say it is timestamp but data type is `date-time` not integer. If is ISO 8601 date format, then mentioned ISO 8601 date time in UTC format.

ramoka178 commented on specification/app/resource-manager/readme.md:
```diff
@@ -65,11 +65,21 @@ input-file:
+      Using the same error response as other APIs.
+  - suppress: LroErrorContent
+    from: Sessions.json
+    reason: |
```
please mention the paths using "where" instead of suppressing for the whole file

yitaopan replied:
Sure, fixed

yitaopan committed:
Fix

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+      "description": "Scale configuration.",
+      "type": "object",
+      "properties": {
+        "maxConcurrentSessions": {
```
Thanks, may I stick to this name since this is already in the codebase and there are other similar names in other apis.

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/Sessions.json:
```diff
@@ -0,0 +1,390 @@
+        "tags": [
+          "ContainerAppsSessionPoolSessions"
+        ],
+        "summary": "Get the properties of a Session.",
```
nit: summary description can be better. like `Get the details of a Session.`

yitaopan replied:
Fixed

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+        "tags": [
+          "ContainerAppsSessionPools"
+        ],
+        "summary": "Get the Session Pools in a given subscription.",
```
update to "in a given resource group of a subscription"

yitaopan replied:
Fixed

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+              "$ref": "#/definitions/SessionPool"
+            }
+          },
+          "201": {
```
HTTP 201 or 202 Created ?
It should be LRO operation in my opinion.

yitaopan replied:
I checked the codebase and other apis, I think this is using 201 Created

TimLovellSmith replied:
201 is correct

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+          "type": "integer",
+          "format": "int32"
+        },
+        "readySessionInstances": {
```
minReadySessionInstances ?

yitaopan replied:
There was a conversation about the naming of this property, and I think the conclusion was readySessionInstances, so I am using this here.

yitaopan replied:
Thanks, there are some changes need to be made in the server codebase in order to meet ARM requirements, which is still in progress, including Patch SessionPool and async Post GenerateSessions.

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+          "description": "Network status for the Sessions.",
+          "type": "string",
+          "enum": [
+            "egressEnabled",
```
Enum value should follow `PascalCasing` pattern like I have seen at multiple places like Status enum etc.

so, it should be `EgressEnabled`

yitaopan replied:
Thanks, you are right, I have changes the enums to be Pascal. For the containerType, I thought we were going with node-lts, but I found this PR https://msazure.visualstudio.com/One/_git/AAPT-Antares-Containerapps/pullrequest/9960874, and decided to make this CustomContainer and PythonLTS now

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+              "type": "string",
+              "readOnly": true
+            },
+            "provisioningState": {
```
this should be under common types. check if you can re-use that

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+      }
+    },
+    "SessionPoolPatchProperties": {
+      "description": "Container App SessionPool patchable properties.",
```
nit: updateable properties ?

yitaopan replied:
Fixed

BlackRider97 commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/examples/SessionPools_CreateOrUpdate.json:
```diff
@@ -0,0 +1,150 @@
+    "sessionPoolName": "testsessionpool",
+    "api-version": "2024-02-02-preview",
+    "sessionPoolEnvelope": {
+      "location": "East US",
```
Isn't it should `eastus` as recommended Azure region name ?
Please check implementation too.

yitaopan replied:
I tried again using az cli --debug, and in the debug message I think it is using similar pattern as "East US"

TimLovellSmith replied:
I think East US should probably be fine, its accepted by most ARM APIs...

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Fix

yitaopan committed:
Remove Sessions Apis

TimLovellSmith commented:
LintDiff needs to pass

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "Create or update succeeded",
```
Only update succeeded should match status 200. Only create succeeded should match 201.

TimLovellSmith replied:
Would be great if you can fix this before this PR ismerged :)

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+          "type": "object",
+          "properties": {
+            "scaleConfiguration": {
+              "description": "The scale configuration of the Session Pool.",
```
A nitpick about doc style - I don't think you should be capitalizing session pool as 'Session Pool' or 'SessionPool' everywhere - it is not a proper noun. Please aim for a clear style of writing.

yitaopan replied:
Hi @TimLovellSmith thanks, I have updated again according to your comments, LintDiff is passed, and I have updated all the "Session Pool" to simple "session pool".
Could you please take another look thanks

yitaopan committed:
fix

yitaopan committed:
fix

yitaopan committed:
fix

yitaopan committed:
fix

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+      }
+    },
+    "SessionPoolCollection": {
+      "description": "Session pool collection ARM resource.",
```
>ARM [](http://example.com/codeflow?start=46&length=3)

Please change it to not mention 'ARM'. You can use the term 'Azure'.

ARM is implementation detail, not something we need to doc.

yitaopan replied:
Fixed

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+            },
+            "poolManagementEndpoint": {
+              "description": "The endpoint to manage the pool.",
+              "type": "string",
```
>   "type": "string", [](http://example.com/codeflow?start=11&length=20)

Is this `"format": "uri"`
?

yitaopan replied:
Thanks, make sense, yes it should be uri, fixed

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+          "x-ms-secret": true
+        }
+      }
+    },
```
I assume you need to model secrets like this because you're providing a general solution for people who are not doing things the cred free way?

yitaopan replied:
Yes, and this follows an existing pattern of another API for container app

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "Secret Name.",
```
>description [](http://example.com/codeflow?start=11&length=11)

Any pattern restriction on secret name?

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+      }
+    },
+    "SessionRegistryCredentials": {
+      "description": "Session pool Private Registry Credentials.",
```
>Private Registry Credentials [](http://example.com/codeflow?start=35&length=28)

nit, can avoid some excessive capitalization in docs

yitaopan replied:
Fixed, went through the file and fixed some other similar docs, too

TimLovellSmith commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+        },
+        "nextLink": {
+          "description": "Link to next page of resources.",
+          "type": "string",
```
`"format": "uri"` should make sense here

yitaopan committed:
fix

yitaopan commented on specification/app/resource-manager/Microsoft.App/preview/2024-02-02-preview/SessionPools.json:
```diff
@@ -0,0 +1,702 @@
+            }
+          },
+          "201": {
+            "description": "Session pool create or update has been started.",
```
Thanks, fixed

TimLovellSmith commented:
Overall looked good to me but I had some worries about the suppression of LRO error linter warning.

If your LRO error response is the same as common definitions, you should just refactor to referencing the common definitions.

If your LRO error response _different_, we should _consider_ swtiching your swagger to using the common definition -- especially if its just going to be adding a few more properties that you aren't using yet.

In some ways, the easiest way to tell, is really just to do the refactor and see what breaking change errors would be the result.

yitaopan commented:
> Overall looked good to me but I had some worries about the suppression of LRO error linter warning.
> 
> If your LRO error response is the same as common definitions, you should just refactor to referencing the common definitions.
> 
> If your LRO error response _different_, we should _consider_ swtiching your swagger to using the common definition -- especially if its just going to be adding a few more properties that you aren't using yet.
> 
> In some ways, the easiest way to tell, is really just to do the refactor and see what breaking change errors would be the result. And remove the suppression.

Thanks, we are aware of this issue, and this Error response is not only used in this Apis but also all the other apis in Microsoft.App, we can see if we can solve this all together in another PR.

yitaopan commented:
/pr RequestMerge
